classes/submissionFile/Repository.php

```diff
@@ -422,7 +424,105 @@ public function copy(SubmissionFile $submissionFile, int $toFileStage, ?int $rev
+                    [
+                        Notification::NOTIFICATION_TYPE_PENDING_INTERNAL_REVISIONS,
+                        Notification::NOTIFICATION_TYPE_PENDING_EXTERNAL_REVISIONS
+                    ],
+                    $authorUserIds,
+                    Application::ASSOC_TYPE_SUBMISSION,
+                    $submissionFile->getData('submissionId')
+                );
+                break;
+
+            case SubmissionFile::SUBMISSION_FILE_COPYEDIT:
+                $notificationMgr->updateNotification(
+                    Application::get()->getRequest(),
+                    [
+                        Notification::NOTIFICATION_TYPE_ASSIGN_COPYEDITOR,
+                        Notification::NOTIFICATION_TYPE_AWAITING_COPYEDITS
+                    ],
+                    null,
+                    Application::ASSOC_TYPE_SUBMISSION,
+                    $submissionFile->getData('submissionId')
+                );
+                break;
+        }
+
+        // Get all revision file ids before they are deleted
+        $revisions = $this->getRevisions($submissionFile->getId());
+
+        // Get the review round before review round files are deleted
+        if ($submissionFile->getData('fileStage') === SubmissionFile::SUBMISSION_FILE_REVIEW_REVISION) {
+            $reviewRoundDao = DAORegistry::getDAO('ReviewRoundDAO'); /** @var ReviewRoundDAO $reviewRoundDao */
+            $reviewRound = $reviewRoundDao->getBySubmissionFileId($submissionFile->getId());
+        }
+
+
         $this->dao->delete($submissionFile);
+
+        // Delete all files that are not referenced by other submission files
+        foreach ($revisions as $revision) {
+            $countFileShares = $this
+                ->getCollector()
+                ->filterByFileIds([$revision->fileId])
+                ->includeDependentFiles(true)
+                ->getCount();
+            if (!$countFileShares) {
+                Services::get('file')->delete($revision->fileId);
+            }
+        }
+
+        // Update the review round status after deletion
+        if ($submissionFile->getData('fileStage') === SubmissionFile::SUBMISSION_FILE_REVIEW_REVISION) {
+            $reviewRoundDao->updateStatus($reviewRound);
+        }
+
+        // Log the deletion
+        $user = Application::get()->getRequest()->getUser();
+        SubmissionFileLog::logEvent(
+            Application::get()->getRequest(),
+            $submissionFile,
+            SubmissionFileEventLogEntry::SUBMISSION_LOG_FILE_DELETE,
+            'submission.event.fileDeleted',
+            [
+                'fileStage' => $submissionFile->getData('fileStage'),
+                'sourceSubmissionFileId' => $submissionFile->getData('sourceSubmissionFileId'),
+                'submissionFileId' => $submissionFile->getId(),
+                'submissionId' => $submissionFile->getData('submissionId'),
+                'username' => $user ? $user->getUsername() : null,
```

Comment:
Just a reminder that we can start using null safe operator from 3.4 as php < 8.0 isn't supported